presto-main/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleNodeSelector.java

```diff
@@ -132,17 +149,32 @@ public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTas
             InternalNode chosenNode = null;
             int min = Integer.MAX_VALUE;
 
-            for (InternalNode node : candidateNodes) {
+            for (int i = 0; i < candidateNodes.size(); i++) {
+                InternalNode node = candidateNodes.get(i);
                 int totalSplitCount = assignmentStats.getTotalSplitCount(node);
+
+                // choose the preferred node first as long as they're not busy
+                if (preferredNodeCount.isPresent() && i < preferredNodeCount.getAsInt() && totalSplitCount < maxSplitsPerNode) {
+                    chosenNode = node;
+                    break;
+                }
+                // fallback to choosing the least busy nodes
                 if (totalSplitCount < min && totalSplitCount < maxSplitsPerNode) {
                     chosenNode = node;
                     min = totalSplitCount;
                 }
             }
             if (chosenNode == null) {
                 // min is guaranteed to be MAX_VALUE at this line
-                for (InternalNode node : candidateNodes) {
```

Comment:
Could you merge the for loop with the above one into a helper? They good almost the same.